lua/neorg/modules/core/latex/renderer/module.lua

```diff
@@ -108,116 +251,254 @@ module.public = {
+        -- position of the images
+        for _, limage in pairs(images) do
+            local range = limage.range
+
+            local ext_opts = {
+                end_col = range[4],
+                strict = false,
+                invalidate = true,
+                undo_restore = false,
+                id = limage.extmark_id, -- if it exists, update it, else this is nil so it will create a new one
+            }
+
+            if conceal_on and range[1] ~= cursor_row - 1 then
+                local image = limage.image
+                local predicted_image_dimensions =
+                    module.private.image_api.image_size(image, { height = module.config.public.scale })
+                ext_opts.virt_text = { { (" "):rep(predicted_image_dimensions.width) } }
+                ext_opts.conceal = ""
+                ext_opts.virt_text_pos = "inline"
             end
+
+            limage.extmark_id =
+                vim.api.nvim_buf_set_extmark(buffer, module.private.extmark_ns, range[1], range[2], ext_opts)
+        end
+
+        for key, limage in pairs(images) do
+            local range = limage.range
+            if conceal_on and range[1] == cursor_row - 1 then
+                table.insert(module.private.cleared_at_cursor, key)
+                module.private.image_api.clear({ limage })
+                goto continue
+            end
+            module.private.image_api.render({ limage })
+            ::continue::
         end
     end,
 }
 
-module.config.public = {
-    -- When true, images of rendered LaTeX will cover the source LaTeX they were produced from
-    conceal = true,
+local running_proc = nil
+local render_timer = nil
+local function render_latex()
+    local buf = vim.api.nvim_get_current_buf()
+    if not module.private.do_render then
+        if render_timer then
+            render_timer:stop()
+            render_timer:close()
+            render_timer = nil
+        end
+        return
+    end
 
-    -- "Dots Per Inch" increasing this value will result in crisper images at the expense of
-    -- performance
-    dpi = 350,
+    if not render_timer then
+        render_timer = vim.uv.new_timer()
+    end
 
-    -- When true, images will render when a `.norg` buffer is entered
-    render_on_enter = false,
+    render_timer:start(module.config.public.debounce_ms, 0, function()
+        render_timer:stop()
+        render_timer:close()
+        render_timer = nil
+
+        if not running_proc then
+            vim.schedule(function()
+                running_proc = nio.run(
+                    function()
+                        module.public.async_latex_renderer(buf)
+                    end,
+                    vim.schedule_wrap(function()
+                        module.public.render_inline_math(module.private.latex_images[buf] or {}, buf)
+                        running_proc = nil
+                    end)
+                )
+            end)
```

Comment:
Is there a reason for the doubly-wrapped schedule + async calls here?

I've been able to replicate the same behaviour with the following snippet:
```suggestion
            running_proc = nio.run(
                function()
                    nio.scheduler()
                    module.public.async_latex_renderer(buf)
                end,
                function()
                    module.public.render_inline_math(module.private.latex_images[buf] or {}, buf)
                    running_proc = nil
                end
            )
```

It doesn't seem to hurt performance nor introduce new bugs. The second function from nio.run is never run from api-fast, so there's no need for a schedule there. `nio.scheduler()` is a cleaner way of doing a whole `vim.schedule` wrap.

Since you're running from a timer callback and the operation itself is async I believe this only introduces visual latency with no extra benefits. Do let me know if I overlooked something :)